engine/rowset/src/main/java/io/deephaven/engine/rowset/impl/sortedranges/SortedRanges.java

```diff
@@ -2087,9 +2087,83 @@ public final OrderedLongSet insertImpl(final SortedRanges other) {
+        // there are two cases, depending on whether we need to merge our last range with
+        // other's first range.
+        final SortedRanges result;
+        final int otherFirstPosToRead; // always the beginning of a range or a singleton (eg, > 0)
+        final int ourFirstPosToWrite;
+        long unpackedLast = unpackedGet(count - 1);
+        if (Math.abs(unpackedLast) + 1 == other.first()) {
+            final boolean weEndInRange = unpackedLast < 0;
+            long otherKeyAtPos1 = -1;
+            if (other.cardinality > 1 && (otherKeyAtPos1 = other.unpackedGet(1)) < 0) {
+                // we are merging a full range from other's front at our end.
+                if (weEndInRange) {
+                    result = ensureCanAppend(count - 1 + other.count - 2, other.last(), writeCheck);
+                } else {
+                    result = ensureCanAppend(count - 1 + other.count - 1, other.last(), writeCheck);
+                }
+                if (result == null) {
+                    return null;
+                }
+                if (weEndInRange) {
+                    result.unpackedSet(count - 1, otherKeyAtPos1);
+                    ourFirstPosToWrite = count;
+                } else {
+                    result.unpackedSet(count, otherKeyAtPos1);
+                    ourFirstPosToWrite = count + 1;
+                }
+                otherFirstPosToRead = 2;
+            } else {
+                // we are merging a single value from other's front at our end.
+                if (weEndInRange) {
+                    result = ensureCanAppend(count - 1 + other.count - 1, other.last(), writeCheck);
+                } else {
+                    result = ensureCanAppend(count - 1 + other.count, other.last(), writeCheck);
+                }
+                if (result == null) {
+                    return null;
+                }
+                if (weEndInRange) {
+                    result.unpackedSet(count - 1, -other.first());
+                    ourFirstPosToWrite = count;
+                } else {
+                    result.unpackedSet(count, -other.first());
+                    ourFirstPosToWrite = count + 1;
+                }
+                otherFirstPosToRead = 1;
+            }
+        } else {
+            // there is no need to merge values from other's front at our end.
+            result = ensureCanAppend(count - 1 + other.count, other.last(), writeCheck);
+            if (result == null) {
+                return null;
+            }
+            otherFirstPosToRead = 0;
+            ourFirstPosToWrite = count;
+        }
+        // copy over the rest.
+        int ourPosToWrite = ourFirstPosToWrite;
+        for (int otherPosToRead = otherFirstPosToRead; otherPosToRead < other.count; ++otherPosToRead) {
+            result.unpackedSet(ourPosToWrite++, other.unpackedGet(otherPosToRead));
+        }
```

Comment:
This looks like a candidate for an array copy, but I assume it's written this way to address different offsets and different internal widths?

Reply:
Yeah, because essentially we are not guaranteed that our two sortedRanges are of the same underlying "type" (eg if they are SortedRangesInt or SortedRangesShort, the unpack part may apply an offset and other fun).
We "could" [TM] optimize to array copy if both are SortedRangesLong (no offset to apply in that case).

Reply:
Ack. This is exactly what I expected.